### PR TITLE
Fix #1518: "custom functions bound to '0' break motion parsing" in a saner way

### LINKED ISSUE #1518: Binding any custom function to "0" in the motion state map causes motions with a "0" in the count argument to fail
# Issue type
- Bug report

# Environment

Emacs version: GNU Emacs 27.2 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.27, cairo version 1.17.4) of 2021-03-26
Operating System: Arch Linux, Linux 5.10.64-1-lts x86_64
Evil version: 1.14.0
Evil installation type: Melpa
Graphical/Terminal: Graphical
Tested in a `make emacs` session (see [CONTRIBUTING.md](https://github.com/emacs-evil/evil/blob/master/CONTRIBUTING.md#bug-reports)): Yes

# Reproduction steps
- Start Emacs
- Evaluate the following lines, which just rename the function bound to 0 while keeping the logic exactly the same:
```
(defalias 'my-beg 'evil-digit-argument-or-evil-beginning-of-line)
(evil-redirect-digit-argument evil-motion-state-map "0" 'my-beg)
```
- Try any motion command with a 0 in the count argument: e.g. `d10j`, `c20l`, `y3040k` and note in the echo area that the interpreted keystrokes are different than expected, meaning the commands don't work as intended.

# Expected behavior
Users should be able to bind custom commands to 0, modifying the behavior of `evil-beginning-of-line` without breaking how motions are read. E.g. I should be able to create a function called `my-beg` that has special behavior, bind it to 0, and still be able to call `d10j` to delete ten lines down.

# Actual behavior
If a custom command is bound to 0, the reading of motions is broken. As an example, when we use `evil-redirect-digit-argument` to bind a command to 0 (just like Evil [does](https://github.com/emacs-evil/evil/blob/82e5becae21b7768ffbb4728bce83dab8751d610/evil-maps.el#L172)), the command `d10j` is interpreted by Emacs as `C-u 62 j`. Therefore, instead of deleting ten lines down, Emacs will jump 62 lines down without modifying the buffer at all.

# Further notes
I discovered this issue after realizing that while `evil-org-mode` was active, any motions with a 0 in the count argument failed, which I documented at Somelauw/evil-org-mode#91. The reason for this issue is as follows:

When an evil operator calls `evil-operator-range` to get the region of text that should operate on, it eventually calls `evil-keypress-parser`, which performs the actual atomic operations of reading and interpreting keystrokes. Typically, the way `evil-keypress-parser` processes a "0" keystroke is to look up whether "0" is bound to `evil-redirect-digit-argument-or-evil-beginning-of-line` *and* if a count argument has already been started. If that's the case, the parser treats "0" just as a digit argument, appending a zero to the end of the current count argument (in effect), and continuing to read for further input.

However, the fact that this takes place as a hardcoded check for a specific function name has a couple implications.

Most notably, if the function bound to 0 in the motion state map has any name other than the one expected, all motion commands with a "0" in the count argument break. When a "0" keystroke is recorded, the branch I described above that depends on the name of the function bound to the key isn't taken, so `evil-keypress parser` doesn't terminate in the expected fashion - the redirected wrapper function just falls through, where it eventually just ends up calling `digit-argument`, with the original operator command getting ignored except for causing some weirdness by modifying the prefix value.

Secondly, `evil-redirect-digit-argument` doesn't actually do anything effective, from what I can tell. In the entire Evil codebase, it's only used to create and bind the wrapper for `evil-beginning-of-line` to 0 - but since this is a motion, any use of it will pass through `evil-read-motion`, and thus `evil-keypress-parser` - and, as stated above, `evil-keypress-parser` already contains logic to determine whether something is a digit argument or another function. In addition, that logic *is* necessary - removing it and trying to just rely on the redirect wrappers to tell if something was a digit argument wouldn't work without heavily rewriting the parser. This means that just using a plain `define-key` to bind `evil-beginning-of-line` to 0 and changing the hardcoded check to compare against `evil-beginning-of-line` instead of its redirected wrapper works just as well from what I can see (although it still has the inherent fragility of the first issue I mentioned).

I believe I've got the beginnings of a PR (it passes all of the tests in evil-tests.el), but I'm new to contributing to Emacs in general and Evil specifically, so I'm aware it may not use canonical methods (e.g. it uses customize-set-variable if `evil-respect-visual-line-mode` is non-nil; the pattern of conditionally setting a custom variable is one I didn't find anywhere else, and don't particularly like).

### PR DESCRIPTION
**Pull request by jmmathena**
This PR improves on #1519, which I made about a week ago. I'm pretty new to Emacs development, so when I made those changes I was imagining some graybeard poweruser with way more experience than me, willfully rebinding digit keys besides "0" to bizarre functions. However, after some reflection I imagine that this is such a rare use case that it isn't really necessary to support it. By way of explanation:

In Vim/Evil, we know that if "0" is the first number we see in a sequence, we should treat it as a motion, because e.g. "d010j" is meaningless - you could just write "d10j". There will never be a case where "0" is *both* the first digit in a sequence *and* has meaning as a number, so when we see a "0" at the beginning of a sequence we're always safe to treat it as a motion instead of part of a count argument.

The variable I added in the *last* PR would allow this behavior with other digits: e.g. "d1" could be a valid and complete command, but you'd also still be able to have "d31j" do as you expect. This comes with the caveat that you *wouldn't* be able to do "d15j" for instance.

In hindsight, this really doesn't make much sense. People do wild stuff with Emacs - I'm sure there's *someone* who would put that facility to good use. However, I imagine it happens infrequently enough that we shouldn't make a special variable just to handle that case, and summarily force people that want to bind functions to "0" to discover and mess with a custom variable, just to accommodate that one weird graybeard.

This PR (instead of checking for a function's inclusion in a list of motions bound to digit keys) simply checks if an entered key is "0" and decides intelligently if its corresponding function should be treated as a digit argument or as a motion. This means that you can bind motions to 0 completely transparently, just like with any other key - no need for weird `evil-redirect-digit-argument` shenanigans (which didn't work), or the additional overhead of adding the function to a custom variable (blech).

In summary, as far as binding stuff to digits goes:

| Strategy | Does it work? | Allows binding to digits besides 0? | 0 motions defined the same as other keybinds? |
| --- | --- | --- | --- |
| `evil-redirect-digit-argument` | :x: | :x: | :x: |
| `evil-digit-bound-motions` | :heavy_check_mark: | :heavy_check_mark: | :x: |
| This PR | :heavy_check_mark: | :x: | :heavy_check_mark: |

### CODE REVIEW

**Comment by tomdl89** [comment]
@jmmathena thanks for your work on this - it's great to see you've given it so much thought. At the risk of annoying one or two (but most likely zero) very keen users, I think we're quick enough that obsoleting so soon after adding is probably a bit too proper. I think we should just remove it and apologize should anyone open an issue complaining. I'll take a look soon.

File: evil-common.el [hunk 1]
```diff
@@ -723,7 +724,7 @@ recursively."
            ((functionp cmd)
             (if (or (memq cmd '(digit-argument negative-argument))
                     (and found-prefix
-                         (memq cmd evil-digit-bound-motions)))
+                         (equal (vconcat seq) [48])))
```
Comment (tomdl89) [inline evil-common.el:727]: Why can't this just be:
```suggestion
                         (string= seq "0")))
```

Comment (jmmathena) [reply evil-common.el:727]: @tomdl89 `evil-extract-count` expects either a string or a vector. If you just use `string=`, it'll fail in cases like [this](https://github.com/emacs-evil/evil/blob/17c635f6e1f538bf4b2c3c276ddd9d4d165a52fb/evil-tests.el#L8617), since `string=` chokes on vectors.

To deal with this, my original solution was to deal with the string and vector cases separately:

```
((functionp cmd)
            (if (or (memq cmd '(digit-argument negative-argument))
                    (and found-prefix
                       (or (equal seq [48])
                            (equal seq "0")))) ;;use equal instead of string= here; if seq is a vector != [48], it'll choke
```

However, since any string can be represented by a vector (the converse is not true - see [M-right], above), we can just always convert `seq` to a vector with `vconcat` and compare it with the character code 48, and avoid having to add an extraneous `or` expression.

Comment (tomdl89) [reply evil-common.el:727]: OK makes sense. I think `(vector ?0)` or `(vconcat "0")` may be more readable than `[48]` but otherwise all good.

**Comment by Somelauw** [comment]
I've just accepted the former change to `evil-digit-bound-motions` in evil-org.
I agree with this PR, though it would temporarily break evil-org again if merged.

**Comment by jmmathena** [comment]
Hey @Somelauw! Appreciate the updates to evil-org-mode. I've just submitted a PR that'll fix evil-org-mode if this PR ends up getting merged. I've been using the commit from this PR for a few weeks now locally, and forked evil-org-mode in the meantime so everything would be copacetic, so I can verify that the change has been working for me. It's a super simple one line edit, but I figured that if I'm breaking it, I should fix it.

**Comment by tomdl89** [comment]
OK I've finally found the time to review this and it's all good - happy to merge. @Somelauw would you like to coordinate merges? If so, I'm UTC+1 so non-nightime hours should be fine. No worries if not.

**Comment by Somelauw** [comment]
Sure, Wednesday 21:30 in UTC+1 would work for me.
Otherwise, Thursday at 20:00.

**Comment by tomdl89** [comment]
@Somelauw great I'll merge Wednesday 21.00 CET
